### PR TITLE
feat(storage): define iter_log method and in-mem trivial implementation

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Define iter_log method as followed
```
fn iter_log(
        &self,
        epoch_range: (u64, u64),
        key_range: TableKeyRange,
        options: ReadLogOptions,
    ) -> impl Future<Output = StorageResult<Self::ChangeLogIter>> + Send + '_;
```

Add trivial implementation for in-memory and sled state store and add unit test.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by MrCroxx — COMMENTED** [review]
Hi, I found some `unimplemented!()`s in this PR. Are they intended and is this PR ready for review now?

**Comment by wenym1** [comment]
> Hi, I found some `unimplemented!()`s in this PR. Are they intended and is this PR ready for review now?

This is expected, since there is no call to the added methods currently. This PR only defines the iter_log method in the StateStoreRead trait. It's split from a large PR https://github.com/risingwavelabs/risingwave/pull/15666

File: src/storage/src/memory.rs [hunk 1]
```diff
@@ -570,8 +572,34 @@ impl<R: RangeKv> StateStoreRead for RangeKvStateStore<R> {
                 to_full_key_range(read_options.table_id, key_range),
             ),
             epoch,
+            true,
         ))
     }
+
+    async fn iter_log(
+        &self,
+        (min_epoch, max_epoch): (u64, u64),
```
Comment (MrCroxx) [inline src/storage/src/memory.rs:581]: 🤣

**Review by MrCroxx — APPROVED** [review]
LGTM

**Pull request merged**